internal/keepers/service.go

```diff
@@ -53,38 +68,27 @@ func newOnDemandUpkeepService(ratio sampleRatio, registry types.Registry, logger
 
 	// start background services
 	s.start()
+
 	return s
 }
 
 var _ upkeepService = (*onDemandUpkeepService)(nil)
 
 func (s *onDemandUpkeepService) SampleUpkeeps(ctx context.Context, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	// get only the active upkeeps from the contract. this should not include
-	// any cancelled upkeeps
-	keys, err := s.registry.GetActiveUpkeepKeys(ctx, "0")
-	if err != nil {
-		return nil, fmt.Errorf("%w: failed to get upkeeps from registry for sampling", err)
-	}
-
-	s.logger.Printf("%d active upkeep keys found in registry", len(keys))
-	if len(keys) == 0 {
-		return []*types.UpkeepResult{}, nil
+	if s.workers == nil {
+		panic("cannot sample upkeeps without runner")
 	}
 
-	// select x upkeeps at random from set
-	keys = s.shuffler.Shuffle(keys)
-	size := s.ratio.OfInt(len(keys))
-
-	s.logger.Printf("%d keys selected by provided ratio %s", size, s.ratio)
-	if size <= 0 {
-		return []*types.UpkeepResult{}, nil
-	}
+	s.eligibleUpkeepKeysLock.Lock()
+	keys := make([]types.UpkeepKey, len(s.eligibleUpkeepKeys))
+	copy(keys, s.eligibleUpkeepKeys)
+	s.eligibleUpkeepKeysLock.Unlock()
 
-	if s.workers == nil {
-		panic("cannot sample upkeeps without runner")
+	if len(keys) == 0 {
+		return nil, nil
 	}
 
-	return s.parallelCheck(ctx, keys[:size], filters...)
+	return s.parallelCheck(ctx, keys, filters...)
```

Comment:
This should no longer be checking keys. The items from the sampling routine `runSamplingUpkeeps` should populate the cache such that the lines above should provide all the results necessary.

Reply:
But filters are part of the SampleUpkeep function call so we can only pre-populate eligible upkeep keys, actual checks should be done based on filters. Or do you mean checking upkeeps for all keys, and then filter final results?

Reply:
You should be able to apply the filters to the list produced by `GetActiveUpkeeps`. From there, just run the key checks.

In the results, you should keep the results from the previous run until the latest run is complete. Both collections should be returned in a `SampleUpkeeps` with filters applied again. This way there is overlap between current and previous head and partial results don't inhibit complete sampling.

Reply:
Done